logicalplan/merge_selects.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+// MergeSelectsOptimizer optimizes a binary expression where
+// one select is a superset of the other select.
+// For example, the expression:
+//
+//	metric{a="b", c="d"} / scalar(metric{a="b"}) becomes:
+//	Filter(c="d", metric{a="b"}) / scalar(metric{a="b"}).
+//
+// The engine can then cache the result of `metric{a="b"}`
+// and apply an additional filter for {c="d"}.
+type MergeSelectsOptimizer struct{}
+
+func (m MergeSelectsOptimizer) Optimize(expr parser.Expr) parser.Expr {
+	heap := make(matcherHeap)
+	extractSelectors(heap, expr)
+	replaceMatchers(heap, &expr)
+
+	return expr
+}
+
+func extractSelectors(selectors matcherHeap, expr parser.Expr) {
+	parser.Inspect(expr, func(node parser.Node, nodes []parser.Node) error {
+		e, ok := node.(*parser.VectorSelector)
+		if !ok {
+			return nil
+		}
+		for _, l := range e.LabelMatchers {
+			if l.Name == labels.MetricName {
+				selectors.add(l.Name, e.LabelMatchers)
+			}
+		}
+		return nil
+	})
+}
+
+func replaceMatchers(selectors matcherHeap, expr *parser.Expr) {
+	traverse(expr, func(node *parser.Expr) {
+		e, ok := (*node).(*parser.VectorSelector)
+		if !ok {
+			return
+		}
+
+		for _, l := range e.LabelMatchers {
+			if l.Name == labels.MetricName {
+				replacement, found := selectors.findReplacement(l.Name, e.LabelMatchers)
+				if found {
+					// All replacements are done on metrics only,
+					// so we can drop the explicit metric name selector.
```

Comment:
`All replacements are done on metrics only` Cannot get this one. Did you mean all replacements are done on labels other than the metric name?

Reply:
So what I wanted to say here is that we only replace selectors that match for metric names. Something like sum({a="b"}) will not be processed with this optimizer. Any suggestions how we can clarify this?

Reply:
Proposed alternative.